core/src/main/java/com/graphhopper/GraphHopper.java

```diff
@@ -529,34 +535,77 @@ public GraphHopper init(GraphHopperConfig ghConfig) {
+            if (name.isEmpty())
+                continue;
+            if (flagEncodersMap.containsKey(name))
+                throw new IllegalArgumentException("Duplicate flag encoder: " + name + " in: " + encoderStr);
+            flagEncodersMap.put(name, encoderStr);
         }
-        Map<String, String> implicitFlagEncoderMap = new LinkedHashMap<>();
+        Map<String, String> flagEncodersFromProfilesMap = new LinkedHashMap<>();
         for (Profile profile : profiles) {
-            emBuilder.add(Subnetwork.create(profile.getName()));
-            if (!flagEncoderMap.containsKey(profile.getVehicle())
-                    // overwrite key in implicit map if turn cost support required
-                    && (!implicitFlagEncoderMap.containsKey(profile.getVehicle()) || profile.isTurnCosts()))
-                implicitFlagEncoderMap.put(profile.getVehicle(), profile.getVehicle() + (profile.isTurnCosts() ? "|turn_costs=true" : ""));
+            // if a profile uses a flag encoder with turn costs make sure we add that flag encoder with turn costs
+            String vehicle = profile.getVehicle().trim();
+            if (!flagEncodersFromProfilesMap.containsKey(vehicle) || profile.isTurnCosts())
+                flagEncodersFromProfilesMap.put(vehicle, vehicle + (profile.isTurnCosts() ? "|turn_costs=true" : ""));
         }
-        flagEncoderMap.putAll(implicitFlagEncoderMap);
-        flagEncoderMap.values().forEach(s -> emBuilder.addIfAbsent(flagEncoderFactory, s));
+        // flag encoders from profiles are only taken into account when they were not given explicitly
+        flagEncodersFromProfilesMap.forEach(flagEncodersMap::putIfAbsent);
 
-        for (String tpStr : encodedValueStr.split(",")) {
-            if (!tpStr.isEmpty()) emBuilder.addIfAbsent(encodedValueFactory, tagParserFactory, tpStr);
-        }
+        List<String> encodedValueStrings = Arrays.stream(encodedValuesStr.split(","))
+                .map(String::trim)
+                .filter(s -> !s.isEmpty())
+                .collect(Collectors.toList());
 
-        return emBuilder.build();
+        EncodingManager.Builder emBuilder = new EncodingManager.Builder();
+        flagEncodersMap.forEach((name, encoderStr) -> emBuilder.add(flagEncoderFactory.createFlagEncoder(name, new PMap(encoderStr))));
+        profiles.forEach(profile -> emBuilder.add(Subnetwork.create(profile.getName())));
+        encodedValueStrings.forEach(s -> emBuilder.add(encodedValueFactory.create(s)));
+        encodingManager = emBuilder.build();
+
+        tagParserBundle = new TagParserBundle();
+        encodedValueStrings.forEach(s -> tagParserBundle.addWayTagParser(tagParserFactory.create(encodingManager, s)));
+
+        // this needs to be in sync with the default EVs added in EncodingManager.Builder#build. ideally I would like to remove
+        // all these defaults and just use the config as the single source of truth
```

Comment:
> I understand what you mean, but IMO something that is required and can't be configured (at least not through the config file) shouldn't be in the config file (?)

Let's see. Currently `RoadAccess` is required mostly because it is required by FastestWeighting. But I think it is debatable if FastestWeighting should even include such a priority thing and we want to switch to using CustomWeighting anyway, so I think we could get rid of this requirement. Then we have the road class and roundabout EVs, which are used for instructions. Roundabout is also used by some vehicle tag parsers (but not all). Maybe someone uses GraphHopper and does not need instructions, so it does not seem to be strictly required either. Another one is RoadEnvironment which is used for Tunnel/Bridge elevation interpolation. Currently we do not require it there, but ignore the interpolation if it is missing (which seems ugly in case we assume the interpolation is always desired).
So we have some encoded values that are required for certain parts of the library, but not necessarily for all use cases.

The other question is whether or not such encoded values should  somehow be added by default. Currently they are not just used as defaults somewhere in the GraphHopper class, but they are added in EncodingManager.Builder.build. So there is no way to create an EncodingManager without these encoded values. I don't like this, because the encoded values might be needed in many cases, but not all. 
The other reason why I don't like the default encoded values is that we already provide a way to specify which encoded values shall be included for a GraphHopper instance. To me this seems easier to understand when this list is definitive. I don't want to first specify some encoded values in the config.yml file and then browse the complicated Java code to find out that some more are added through the backdoor. 

> But I would be fine too if we throw an exception if a required encoded value is missing (otherwise, if we completely ignore this problem then graphhopper would throw an exception too late).

Yes, I fully agree. If some feature needs some encoded value it should error when the encoded value is missing. But if the feature is not used/enabled/etc. the requirement of the encoded value is gone so everything should work without it.

> (Or we remove the default EVs from the EncodingManager only and add them together with the TagParsers here or something?)

Yes. Adding them in EncodingManager.Builder.build is the worst part of the default EVs I think. Because this means that we enforce them on a very low level where it is not clear whether they are required or not. The problem when we not add them on this low level is that we need to add them explicitly when they are needed (in many tests for example). But in some cases we will also not need to add them because we weren't using them. This is one example of what I meant with "How to deal with boilerplate code, especially after above changes?" Still, I think it is the right thing to do and adding them can be reduced to one call of some static helper method or something.

> And if we assume the config to be the single source of truth, would you create a tag parser for every encoded value (and throw an exception if a required EV is missing) or should the tag parsers be specified separately in the config?

Currently we have `graph.encoded_values` and `graph.flag_encoders`. So far `graph.encoded_values` should have really been called `graph.tag_parsers`, because it controlled the list of tag parsers, which in turn specified which encoded values will be created (apart from the default ones). I kept it as before to avoid too many changes in this PR, but maybe we need to think about this again. I use the list of encoded values to create both the encoded values and the tag parsers (this is why there is an EncodedValueFactory again, but this time it is actually being used each time we create a GraphHopper instance). So as it stands all we'd have to do would be adding something like `road_environment` to `graph.encoded_values` and this would add the RoadEnvironment EV to the encoded values and add the tag parser as well. But I think we should get rid of `graph.flag_encoders` and the VehicleTagParsers are a little different, because they require multiple encoded values. This seems to require a bit more thought so I would do it after this PR.

Reply:
Ok. But let's do this after this PR?